pages/_app.tsx

```diff
@@ -131,18 +92,31 @@ const App = ({ apiKey, ...props }: AppProps & AuthProps) => {
     head = <props.Component {...props.pageProps} headOnly />;
   }
 
-  const [token, setToken] = useState<{ token?: string } | undefined>(apiKey ? { token: apiKey } : undefined);
+  const tokenFromHeader = apiKey ? { token: apiKey } : undefined;
+  const tokenFromURL = isTest() ? { token: getReadOnlyParamsFromURL().apiKey } : undefined;
+  const [token, setToken] = useState<{ token: string | null } | undefined>(tokenFromHeader || tokenFromURL);
   useEffect(() => {
     async function fetchToken() {
       const response = await fetch("/api/token");
-      const token = response.ok ? await response.text() : undefined;
+      const token = response.ok ? await response.text() : null;
       setToken({ token });
     }
+
+    async function getTokenFromBrowser() {
+      const token = await Promise.race([
+        new Promise<string>(resolve => listenForAccessToken(resolve)),
+        // if the user is not logged in, the listenForAccessToken callback
+        // is never called, so we add this timeout to continue without a token
+        new Promise<null>(resolve => setTimeout(() => resolve(null), 500)),
+      ]);
```

Comment:
I _guess_ this makes sense but it seems kind of iffy. Is 500ms enough time to always be safe?

Reply:
Yeah, I'd prefer if the browser sent us a "no access token" response. Anyway in my tests the browser responded within a few milliseconds, so 500ms should be enough.